### PR TITLE
Allow nullable types in Flow's firstOrNull extension

### LINKED ISSUE #2229: Flow.firstOrNull should allow nullable types
`Flow.firstOrNull` does not allow nullable types, presumably so that you can differentiate between the flow being empty or not. However, every other `firstOrNull` function such as `List.firstOrNull` does allow nullable types and I see no technical reason why the flow version should not allow nullable types.

This change would be binary compatible since the type argument's nullability isn't even reflected in the bytecode. It would also be source compatible since it only loosens the restrictions.

### PR DESCRIPTION
**Pull request by ansman**
This closes #2229

### CODE REVIEW

**Review by qwwdfsad — COMMENTED** [review]
Thanks!

Could you please also remove `Any` upper bound from `singleOrNull` operator and add a corresponding unit test?

**Comment by ansman** [comment]
Done

File: kotlinx-coroutines-core/common/src/flow/terminal/Reduce.kt [hunk 1]
```diff
@@ -138,10 +138,10 @@ public suspend fun <T : Any> Flow<T>.firstOrNull(): T? {
 }
 
 /**
- *  The terminal operator that returns the first element emitted by the flow matching the given [predicate] and then cancels flow's collection.
+ * The terminal operator that returns the first element emitted by the flow matching the given [predicate] and then cancels flow's collection.
  * Returns `null` if the flow did not contain an element matching the [predicate].
```
Comment (ansman) [inline kotlinx-coroutines-core/common/src/flow/terminal/Reduce.kt:142]: Do we want to comment on what happens if the flow contains a single `null` value or let that be implied?

**Review by qwwdfsad — CHANGES_REQUESTED** [review]
`assertFailsWith<NoSuchElementException> { flowOf<Int?>().singleOrNull() }` doesn't actually fail but returns `null` instead. Added test doesn't pass

**Comment by ansman** [comment]
My bad. I'm unable to run the tests locally because java 6 is needed.

**Comment by LouisCAD** [comment]
@ansman Read the updated `CONTRIBUTING.md` on the main/master branch, it's specified you can make the env variable point to a newer JDK, including JDK11+

**Review by qwwdfsad — APPROVED** [review]
Multiple `null` values are handled incorrectly.
I've created #2287 with your changes and fixed it in https://github.com/Kotlin/kotlinx.coroutines/pull/2287/commits/7d006bbf028e3575469f35ddc50411afc4bb122e to reduce a rountrip time and merge it.

Thanks for raising the issue and your contribution, it's always great to evolve our API based on our users feedback!

**Pull request closed without merging**